packages/utils/src/metrics.ts

```diff
@@ -0,0 +1,80 @@
+export type LabelsGeneric = Record<string, string | number>;
+export type LabelKeys<Labels extends LabelsGeneric> = Extract<keyof Labels, string>;
+export type CollectFn<Labels extends LabelsGeneric> = (metric: Gauge<Labels>) => void;
+
+// Type `{}` is required here as alternative types do not work
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type NoLabels = {};
+
+export interface Gauge<Labels extends LabelsGeneric = NoLabels> {
+  startTimer(): NoLabels extends Labels ? () => number : (labels: Labels) => number;
+  startTimer(labels: NoLabels extends Labels ? undefined : Labels): (labels?: Labels) => number;
+
+  inc: NoLabels extends Labels ? (value?: number) => void : (labels: Labels, value?: number) => void;
+  dec: NoLabels extends Labels ? (value?: number) => void : (labels: Labels, value?: number) => void;
+  set: NoLabels extends Labels ? (value: number) => void : (labels: Labels, value: number) => void;
+
+  collect?(): void;
+}
+
+export interface GaugeExtra<Labels extends LabelsGeneric = NoLabels> extends Gauge<Labels> {
+  addCollect(collectFn: CollectFn<Labels>): void;
+
+  /**
+   * @deprecated Use `addCollect` instead
+   */
+  collect(): void;
+}
+
+export interface Histogram<Labels extends LabelsGeneric = NoLabels> {
+  startTimer(): NoLabels extends Labels ? () => number : (labels: Labels) => number;
+  startTimer(labels: NoLabels extends Labels ? undefined : Labels): (labels?: Labels) => number;
+
+  observe: NoLabels extends Labels ? (value: number) => void : (labels: Labels, value: number) => void;
+
+  reset(): void;
+}
+
+export interface AvgMinMax<Labels extends LabelsGeneric = NoLabels> {
+  addGetValuesFn(getValuesFn: () => number[]): void;
+
+  set: NoLabels extends Labels ? (values: number[]) => void : (labels: Labels, values: number[]) => void;
+}
+
+export interface Counter<Labels extends LabelsGeneric = NoLabels> {
+  inc: NoLabels extends Labels ? (value?: number) => void : (labels: Labels, value?: number) => void;
+}
+
+export type GaugeConfig<Labels extends LabelsGeneric> = {
+  name: string;
+  help: string;
+} & (NoLabels extends Labels ? {labelNames?: never} : {labelNames: [LabelKeys<Labels>, ...LabelKeys<Labels>[]]});
```

Comment:
This type is same as:

```suggestion
} & (NoLabels extends Labels ? {labelNames?: never} : {labelNames: LabelKeys<Labels>[]});
```

Reply:
The behavior is different

`labelNames: LabelKeys<Labels>[]` allows empty arrays
```ts
blockProductionTime: register.histogram<{source: ProducedBlockSource}>({
  // ...
  labelNames: [], // No complaints
}),
```

whereas `labelNames: [LabelKeys<Labels>, ...LabelKeys<Labels>[]]` requires you to actually add the label name
```ts
blockProductionTime: register.histogram<{source: ProducedBlockSource}>({
  // ...
  labelNames: [], // Type '[]' is not assignable to type '["source", ..."source"[]]'.
}),
```

The `...LabelKeys<Labels>[]` part is required for cases were we have multiple labels